tasks/Task.cpp

```diff
@@ -2,48 +2,331 @@
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::GatheringState::Complete:
+                mState.gathering_state = Complete;
+            case rtc::PeerConnection::GatheringState::InProgress:
+                mState.gathering_state = InProgress;
+            case rtc::PeerConnection::GatheringState::New:
+                mState.gathering_state = NewGathering;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onSignalingStateChange(
+        [&](rtc::PeerConnection::SignalingState state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::SignalingState::HaveLocalOffer:
+                mState.signaling_state = HaveLocalOffer;
+            case rtc::PeerConnection::SignalingState::HaveLocalPranswer:
+                mState.signaling_state = HaveLocalPranswer;
+            case rtc::PeerConnection::SignalingState::HaveRemoteOffer:
+                mState.signaling_state = HaveRemoteOffer;
+            case rtc::PeerConnection::SignalingState::HaveRemotePranswer:
+                mState.signaling_state = HaveRemotePranswer;
+            case rtc::PeerConnection::SignalingState::Stable:
+                mState.signaling_state = Stable;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onLocalDescription(
+        [&, protocol, remote_peer_id](rtc::Description description)
+        {
+            Json::Value message;
+            Json::FastWriter fast;
+            message["protocol"] = protocol;
+            message["to"] = remote_peer_id;
+            message["actiontype"] = description.typeString();
+            message["description"] = string(description);
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+                ws->send(fast.write(message));
+        });
+
+    peer_connection->onLocalCandidate(
+        [&, protocol, remote_peer_id](rtc::Candidate candidate)
+        {
+            Json::Value message;
+            Json::FastWriter fast;
+            message["protocol"] = protocol;
+            message["to"] = remote_peer_id;
+            message["actiontype"] = "candidate";
+            message["candidate"] = string(candidate);
+            message["mid"] = candidate.mid();
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+                ws->send(fast.write(message));
+        });
+
+    return peer_connection;
+}
+
+void Task::configurePeerDataChannel(string const &remote_peer_id)
 {
+    mPeerConnection->onDataChannel(
+        [&](shared_ptr<rtc::DataChannel> data_channel)
+        {
+            LOG_INFO_S << "DataChannel from " << remote_peer_id
+                       << " received with label \"" << data_channel->label() << "\""
+                       << std::endl;
+
+            mDataChannel = data_channel;
+            data_channel->onOpen([&, wdc = make_weak_ptr(data_channel)]() {});
```

Comment:
I think you should be filling something in the status structure when this happens ...